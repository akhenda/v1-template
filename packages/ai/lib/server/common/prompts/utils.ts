import type { AnyValue } from '@repo/types';

/**
 * Splits the input prose into an array of trimmed chunks based on the `---CHUNK---` delimiter.
 *
 * @param prose - The string containing prose with `---CHUNK---` delimiters.
 * @returns An array of prose chunks, each trimmed of surrounding whitespace.
 */
export function getPromptChunksFromProse(prose: string) {
  return prose
    .split('---CHUNK---')
    .map((c) => c.trim())
    .filter(Boolean);
}

/**
 * Joins an array of prose chunks into a single string, separated by a blank line in between.
 *
 * @param chunks - An array of strings, each containing a chunk of prose.
 * @returns A single string containing the joined prose chunks.
 */
export function getPromptFromChunks(chunks: string[]) {
  return chunks.join('\n\n');
}

/**
 * Creates a prompt string from a tagged template literal, splitting it into chunks at
 * `---CHUNK---` delimiters. Each chunk is trimmed of surrounding whitespace, and any
 * empty chunks are removed.
 *
 * @param literals - A tagged template literal containing the prompt text.
 * @param placeholders - Zero or more values to be interpolated into the prompt text.
 * @returns A single string containing the joined and trimmed prompt chunks.
 */
export function chunkedPrompt(literals: TemplateStringsArray, ...placeholders: AnyValue[]) {
  const raw = String.raw(literals, ...placeholders);

  return raw
    .split('---CHUNK---')
    .map((c) => c.trim())
    .filter(Boolean)
    .join('\n\n')
    .trim();
}

/**
 * Transforms a markdown prompt file into a flat string with `---CHUNK---` before each H3 section.
 * Returns a clean string ready to pass to a prompt function.
 */
export function chunkMarkdownPrompt(raw: string, injectResumeSchema = false) {
  const transformed = raw
    .split('\n---\n')
    .map((c) => c.trim())
    .filter(Boolean)
    .join('\n\n\n')
    .trim();

  if (injectResumeSchema) return transformed.replace('___REPLACE_WITH_SCHEMA___', '');

  return transformed;
}

/**
 * Sanitizes interpolated input values like role titles for use in prompts.
 * Escapes backslashes, quotes, and line breaks.
 */
export function sanitizePromptInput(value: string): string {
  return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, ' ').replace(/\r/g, '');
}
