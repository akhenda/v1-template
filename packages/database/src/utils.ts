import * as R from 'remeda';

import { log } from '@repo/observability/log';

export const noop = R.doNothing;
export const logger = log;
