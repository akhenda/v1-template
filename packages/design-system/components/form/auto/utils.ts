import { isValidPhoneNumber, parsePhoneNumberWithError } from 'libphonenumber-js';
import type { DefaultValues, FieldValues } from 'react-hook-form';
import { z } from 'zod';

import type { AnyValue } from './types';

export function getDefaultValueInZodStack(schema: z.ZodTypeAny): AnyValue {
  if (schema instanceof z.ZodDefault) return schema._def.defaultValue();
  if (schema instanceof z.ZodEffects) return getDefaultValueInZodStack(schema.innerType());

  return undefined;
}

export function getDefaultValues<Schema extends z.ZodObject<AnyValue>, Inputs extends FieldValues>(
  schema: Schema,
): DefaultValues<Inputs> {
  const objectSchema = schema instanceof z.ZodEffects ? schema.innerType() : schema;
  const shape = objectSchema.shape;
  const defaultValues = {} as DefaultValues<Inputs>;

  for (const [key, field] of Object.entries(shape)) {
    const defaultValue = getDefaultValueInZodStack(field as z.ZodTypeAny);

    if (defaultValue) {
      defaultValues[key as keyof DefaultValues<Inputs>] = defaultValue;
    }
  }

  return defaultValues;
}

export function phoneNumberSchema(schema: z.ZodString) {
  return schema
    .refine(
      isValidPhoneNumber,
      'Please specify a valid phone number (include the international prefix).',
    )
    .transform((value) => parsePhoneNumberWithError(value).number.toString());
}
