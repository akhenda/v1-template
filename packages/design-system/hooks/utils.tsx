import { createContext, useContext } from 'react';

/**
 * Creates a new context and returns a hook and provider for accessing the
 * context value.
 *
 * @template ContextType - The type of the context value.
 * @param {string} name - The name of the context.
 * @returns {Array} `[useContext, ContextProvider]` - The hook and provider for
 * accessing the context value.
 * @throws {Error} - If the hook is used outside of the provider.
 */
export function createCtx<ContextType>(name: string) {
  const newContext = createContext<ContextType | null>(null);

  /**
   * Returns the context value from the given context.
   *
   * @template T - The type of the context value.
   * @returns {T & ContextType} - The context value.
   * @throws {Error} - If the hook is used outside of the provider.
   */
  function useCtx<T>() {
    const ctx = useContext(newContext);

    if (!ctx) throw new Error(`Hook used outside of the provider: ${name}`);

    return ctx as T & ContextType;
  }

  return [useCtx, newContext.Provider] as const;
}

/**
 * Creates a context with the specified type and returns a hook to access the
 * context value.
 *
 * @param {string} name - The name of the context.
 * @return {Array} `[useContext, Context]` - The hook and provider for
 * accessing the context value.
 */
export function createCtxFull<ContextType>(name: string) {
  const newContext = createContext<ContextType | null>(null);

  /**
   * Returns the context value from the given context.
   *
   * @template T - The type of the context value.
   * @returns {T & ContextType} - The context value.
   * @throws {Error} - If the hook is used outside of the provider.
   */
  function useCtx() {
    const ctx = useContext(newContext);

    if (!ctx) throw new Error(`Hook used outside of the provider: ${name}`);

    return ctx;
  }

  return [useCtx, newContext] as const;
}
