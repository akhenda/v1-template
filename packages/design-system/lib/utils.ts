import { clsx } from 'clsx';
import type { ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

/**
 * A utility function that merges `tailwind-merge` and `clsx`
 * for better compatibility with NativeWind.
 *
 * @param inputs - A list of class names, or functions that return class names.
 * @returns A merged class name.
 */
export const cn = (...inputs: ClassValue[]): string => twMerge(clsx(inputs));

export function capitalize(str: string) {
  return  str.charAt(0).toUpperCase() + str.slice(1);
}

export function getAvatarUrl(seed: string) {
  return `https://api.dicebear.com/9.x/glass/svg?seed=${seed}`;
}
