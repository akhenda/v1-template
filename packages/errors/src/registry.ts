export const ERROR_REGISTRY = {
  // HTTP Request Errors
  RequestError: { code: 500 },
  NotFoundError: { code: 404 },
  BadRequestError: { code: 400 },
  UnauthorizedError: { code: 401 },
  ForbiddenError: { code: 403 },
  ConflictError: { code: 409 },
  TooManyRequestsError: { code: 429 },
  InternalServerError: { code: 500 },
  RequestTimeoutError: { code: 408 },

  // ATS/Domain-Specific Errors
  ATSError: { code: 500 },
  InvalidATSJSONPointerError: { code: 400 },
  InvalidATSSuggestionValueError: { code: 400 },
} as const;

export type ErrorName = keyof typeof ERROR_REGISTRY;
export type ErrorCode = (typeof ERROR_REGISTRY)[ErrorName]['code'];

export function getErrorCode(errorName: ErrorName): ErrorCode {
  return ERROR_REGISTRY[errorName].code;
}
