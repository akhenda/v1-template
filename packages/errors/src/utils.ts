// Utility functions for creating error-specific helpers

import type { AnyValue } from '@repo/types';

export interface ErrorConstructor<T> {
  new (message: AnyValue): T;
}

/**
 * Creates an assertion function for a specific error type
 */
export function createAssertionFn<T extends Error & { cause?: unknown }>(
  ErrorClass: ErrorConstructor<T>,
) {
  return (condition: unknown, message: string, cause?: unknown): asserts condition => {
    if (!condition) {
      const err = new ErrorClass(message);

      if (cause) err.cause = cause;

      throw err;
    }
  };
}

/**
 * Creates a type guard function for a specific error type
 */
export function createTypeGuard<T extends Error>(ErrorClass: ErrorConstructor<T>) {
  return (err: unknown): err is T => err instanceof ErrorClass;
}

/**
 * Creates a simpler assertion function for basic error throwing
 */
export function createSimpleAssertionFn<T extends Error>(ErrorClass: ErrorConstructor<T>) {
  return (condition: unknown, message?: string): asserts condition => {
    if (!condition) throw new ErrorClass(message || 'Assertion failed');
  };
}

/**
 * Utility to check if an error is of a specific type
 */
export function isErrorInstance<T extends Error>(
  err: unknown,
  ErrorClass: ErrorConstructor<T>,
): err is T {
  return err instanceof ErrorClass;
}

/**
 * Creates consistent error utilities for any error type
 */
export function createErrorUtilities<T extends Error>(ErrorClass: ErrorConstructor<T>) {
  return {
    is: createTypeGuard(ErrorClass),
    assert: createAssertionFn(ErrorClass),
    assertSimple: createSimpleAssertionFn(ErrorClass),
    isInstance: (err: unknown) => isErrorInstance(err, ErrorClass),
  };
}
