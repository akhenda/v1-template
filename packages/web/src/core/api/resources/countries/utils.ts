export type RESTCountriesAPIErrorResponse = { message: string; status: number };
