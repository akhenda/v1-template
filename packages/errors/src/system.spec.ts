import { describe, expect, it } from 'vitest';

import { ERROR_REGISTRY, getErrorCode } from './registry';
import { createErrorUtilities } from './utils';

// Test Error classes
class TestError extends Error {}
class AnotherTestError extends Error {}

describe('Error Registry', () => {
  it('should have a comprehensive list of error codes', () => {
    expect(ERROR_REGISTRY).toMatchSnapshot();
  });

  it('should return the correct error code for a given error name', () => {
    expect(getErrorCode('NotFoundError')).toBe(404);
    expect(getErrorCode('BadRequestError')).toBe(400);
    expect(getErrorCode('InvalidATSJSONPointerError')).toBe(400);
    expect(getErrorCode('ATSError')).toBe(500);
  });
});

describe('Shared Error Utilities', () => {
  const { is, assert, assertSimple, isInstance } = createErrorUtilities(TestError);

  it('should create a working type guard', () => {
    const err = new TestError('test error');
    expect(is(err)).toBe(true);
    expect(is(new AnotherTestError('another error'))).toBe(false);
  });

  it('should create a working assertion function', () => {
    const message = 'Assertion failed';
    expect(() => assert(false, message)).toThrow(new TestError(message));
    expect(() => assert(true, message)).not.toThrow();
  });

  it('should create an assertion function that includes cause', () => {
    const cause = new Error('root cause');
    try {
      // @ts-ignore: Testing with a cause
      assert(false, 'with cause', cause);
    } catch (e: unknown) {
      expect((e as TestError).cause).toBe(cause);
    }
  });

  it('should create a simple assertion function', () => {
    expect(() => assertSimple(false, 'Simple assertion')).toThrow(
      new TestError('Simple assertion'),
    );
    expect(() => assertSimple(true, 'Simple assertion')).not.toThrow();
  });

  it('should create a working instance checker', () => {
    const err = new TestError('test error');
    expect(isInstance(err)).toBe(true);
    expect(isInstance(new AnotherTestError('another error'))).toBe(false);
  });
});
