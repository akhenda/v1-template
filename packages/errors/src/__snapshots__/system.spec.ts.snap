// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Error Registry > should have a comprehensive list of error codes 1`] = `
{
  "ATSError": {
    "code": 500,
  },
  "BadRequestError": {
    "code": 400,
  },
  "ConflictError": {
    "code": 409,
  },
  "ForbiddenError": {
    "code": 403,
  },
  "InternalServerError": {
    "code": 500,
  },
  "InvalidATSJSONPointerError": {
    "code": 400,
  },
  "InvalidATSSuggestionValueError": {
    "code": 400,
  },
  "NotFoundError": {
    "code": 404,
  },
  "RequestError": {
    "code": 500,
  },
  "RequestTimeoutError": {
    "code": 408,
  },
  "TooManyRequestsError": {
    "code": 429,
  },
  "UnauthorizedError": {
    "code": 401,
  },
}
`;
